### ISSUE TITLE
Need to restart crowsnest service every time for camera to show

### ISSUE THREAD

**Issue by cpasoft** [mainsail-crew/crowsnest#117]
### What happened

On boot, the camera is not working (black screen with FPS:00).
Go to services control, stop, start and working... 

I've uploaded the log.
Version: v4.0.2-0-gb0c011b
[crowsnest_fail on boot.log](https://github.com/mainsail-crew/crowsnest/files/11568948/crowsnest_fail.on.boot.log)

Downgrade to 3.0.7.3 is working well, here is the log.
[crowsnest_version v3.0.7-3-g20ed6a8.log](https://github.com/mainsail-crew/crowsnest/files/11568952/crowsnest_version.v3.0.7-3-g20ed6a8.log)

@KwadFan upload a test branch with a sleep in the init stream, but still not working.
Here is the log:
[at boot _ not working.txt](https://github.com/mainsail-crew/crowsnest/files/11568960/at.boot._.not.working.txt)

But again, if I stop and re-run the service, then It starts working. Here is the log:
[stop & start _ working.txt](https://github.com/mainsail-crew/crowsnest/files/11568962/stop.start._.working.txt)

comparing the two, there is a line that only appears in the boot, when it does not work.
This line is:
[05/25/23 22:01:14] crowsnest: DEBUG: ustreamer [cam 1]: [warn] getaddrinfo: address family for nodename not supported

Don't know if this is important or not...

### What did you expect to happen

Working right away on boot

### How to reproduce

-

**Comment by LoganMD** [comment]
Second, my camera doesn't work anymore

**Comment by meteyou** [comment]
@LoganMD did you read any announcements or other issues?

**Comment by Senthiele** [comment]
i third my cam is down now

**Comment by ichbinviktor** [comment]
Same here,
i configurated another MJPGStreamer for Rpicam, and crowsnest v4 for the 2nd USB cam.
After every reboot, the video from crowsnest always shows 00 FPS and a black frame, which resumes after manually "save & restart" the crowsnest.conf to make it work.
After configuring the raspberry-pi-cam according to [https://crowsnest.mainsail.xyz/faq/upgrade-from-v3-to-v4](url), both cams can be displayed after boot.
Btw, this instruction for RPI-cam with crowsnest v4 is not easy to find... That's why i used another streamer at the beginning.
[manually-restart.txt](https://github.com/mainsail-crew/crowsnest/files/11628884/manually-restart.txt)

**Comment by DomiStyle** [comment]
Adding a simple 5s delay to the Crowsnest start solves this:

`sudo nano /etc/systemd/system/crowsnest.service`

Add `ExecStartPre=/bin/sleep 5` to the `[Service]` section.

**Comment by cpasoft** [comment]
I  confirm that the workaround of adding a pause at startup has worked...
Thanks @DomiStyle

**Comment by cpasoft** [comment]
@KwadFan , what do you want to do with this issue??...

**Comment by KwadFan** [comment]
Currently working on a rewrite of some crucial parts. So, please be patient will take that into account. 

Regards Kwad

**Comment by cpasoft** [comment]
Thank you Kwad... Let us know if you need help with a test or something...

Regards.

**Comment by gerbenvandekraats** [comment]
Some issue, workaround fixed it for now :)

**Comment by mryel00** [comment]
Could it be that you are all using Ubuntu as your OS? Just realized this after #192

**Comment by mryel00** [comment]
@lin-ycv and everyone else having this issue. What image did you use? I'm trying to reproduce this but it seems like it might be connected to an old Linux kernel 4.9.191 or the specific LTS version of 20.04.6.
I currently have 3 logs about this issue. All 3 of them have those two versions I mentioned.
So I would really appreciate if someone could link me to that image or maybe some people could give me some more logs to get more data on this issue.
If I cannot reproduce it, I won't be able to fix it. As the only option of fixing this, without further information, is to just add a sleep for every Ubuntu user, therefore I most likely will not add such a delay for now.

**Comment by mryel00** [comment]
Or could it be that you are all using the Flsun SpeederPad?

**Comment by lin-ycv** [comment]
I'm on speederpad. I'm away from the printers for a week, will check for `webcam.txt` when I'm back in the shop.

**Comment by mryel00** [comment]
> I'm on speederpad.

That confirms my first assumption😄 
> I'm away from the printers for a week, will check for `webcam.txt` when I'm back in the shop.

Also I just got new infos, You might have to look under `~/klipper_config`. So have a look at both folders please.
Looking forward to your answer 😄

**Comment by mryel00** [comment]
@cpasoft Have you already tried to run it on a different port than 8080? Please try out 8089 if possible and have a look at your log, if the same issue appears. Ofc without the fix within the service file.
If the error then still appears, it's about the startup behaviour of crowsnest and the SpeederPad and I will add a sleep specifically for SpeederPads in the service file. This fix would only apply for new installations and I would like to avoid this one.
If it doesn't appear then anymore, then there might be still something else blocking it at startup.

**Comment by cpasoft** [comment]
For sure, I'll test this... Let me 15 minutes...

**Comment by cpasoft** [comment]
> @cpasoft Have you already tried to run it on a different port than 8080? Please try out 8089 if possible and have a look at your log, if the same issue appears. Ofc without the fix within the service file. If the error then still appears, it's about the startup behaviour of crowsnest and the SpeederPad and I will add a sleep specifically for SpeederPads in the service file. This fix would only apply for new installations and I would like to avoid this one. If it doesn't appear then anymore, then there might be still something else blocking it at startup.

Same behaivour on the 8089 port...
Here is the log...
[crowsnest.log](https://github.com/mainsail-crew/crowsnest/files/13284030/crowsnest.log)

If you want me to try/test anything else, just let me know.

**Comment by mryel00** [comment]
Ok, thank you for the quick test.
Then I will add the sleep inside the service file for the SpeederPad. Like written above, this will only fix it for fresh installations, but you can manually copy the service file to the directory or just reinstall crowsnest as soon as this will be implemented.

**Comment by cpasoft** [comment]
Thanks, this workaround has worked well for me for a while, so I think it is a good solution. 
I'd like to know what's causing this, as it worked perfectly with previous versions, but who knows?

**Comment by mryel00** [comment]
> If you want me to try/test anything else, just let me know.

@cpasoft As you already asked for it I would kindly ask you to test a specific change for me. I don't have uploaded something for it yet, so you would have to add these changes manually.
Please remove the workaround in the service file. Then add following code:
```bash
if [[ "$(awk '{print $1}' /proc/uptime)" -lt 60 ]]; then
    sleep 5
fi
```
Before the `init_logging` in `~/crowsnest/crowsnest` (at line 67 to be specific)
https://github.com/mainsail-crew/crowsnest/blob/4050a3bcf846f2c97142f2569bdf3c9cabbd6f54/crowsnest#L66-L68

This should add a sleep of 5s but only at the first startup. This is only a small test for the real change later.

**Comment by cpasoft** [comment]
> @cpasoft As you already asked for it I would kindly ask you to test a specific change for me.
For sure, let me 15 minutes and I'll let you know how it works

**Comment by cpasoft** [comment]
Modify the file as shown here:
![imagen](https://github.com/mainsail-crew/crowsnest/assets/29692062/fd033650-0649-409b-b5e2-37bd71faf8c8)
 Remove the workaround from the service file, reboot and...
Not working... same results...

Modify the sleep time up to 8 seconds, reboot and, again, nothing.

Then, I've debug the code and I've fixed it to work to:
![imagen](https://github.com/mainsail-crew/crowsnest/assets/29692062/1a1fcf0f-1fe0-4dec-aa79-4583069322da)

And now, it finally working...

Here is the code...
```
if (( $(awk '{print $1}' /proc/uptime | cut -d '.' -f 1) < 160 )); then
    sleep 5
fi
```

It was failing because the decimal point.

**Comment by mryel00** [comment]
I should have tested it before sending 😅
I'm not really a bash dev, so I'm very thankful for debugging and fixing it ❤️ 
What about the startup time? Is 60 fine or should I raise it? I don't know how fast the SpeederPad is, but I would assume 60s should be enough?

**Comment by cpasoft** [comment]
No problem. I like to be helpful. 

I think this value is perfect. I've tested several times and work well.

**Comment by mryel00** [comment]
Just to be sure that there is no misunderstanding 😅
So 60 works well or 160?

**Comment by cpasoft** [comment]
60 is fine, sorry...
You could put 100 if you want to feel more confident.

**Comment by mryel00** [comment]
To give some explanation for that PR and how to correctly use it.
After the PR will be merged into the master branch with the next update (v4.0.6), there will be a new shell parameter. That shell parameter can be used inside the environment file `~/printer_data/systemd/crowsnest.env`.
Just add `-s` at the end of the file. It should then look similar to:
`CROWSNEST_ARGS="-c /home/pi/printer_data/config/crowsnest.conf -s"`
Make sure it's before the last `"`.
This will add a 5s delay to the crowsnest start, but only at the intial boot of the SpeederPad/Linux device. You can also increase/decrease the sleep timer by adding a number behind the `-s`, e.g.:
`CROWSNEST_ARGS="-c /home/pi/printer_data/config/crowsnest.conf -s 10"`
After adding that parameter, you can remove the `ExecStartPre=/bin/sleep 5` from your service file at `/etc/systemd/system/crowsnest.service`.

P.S. If it's shown as merged, it's not released. Wait until you see an update in your update manager. This issue should get automatically closed at that time too.

**Comment by mryel00** [comment]
This parameter is now integrated with [v4.1.0](https://github.com/mainsail-crew/crowsnest/releases/tag/v4.1.0). This will be automatically added on new installations.

**Comment by gralves2** [comment]
I am having exactly the same issue with a C270 camera on RPi 4 on crownest 4.1.3. This workaround worked for me (putting sleep on the Service section).

**Comment by mryel00** [comment]
@gralves2 Did you read the last two comments from me on this post? No need to manually edit the service file....
I will hide that workaround that people hopefully read the rest of the post with the intended way of fixing it.

Also this shouldn't happen on a pi4. The reason why this happens is some problem with the OS not reporting the network startup correctly. This should only happen on the Flsun Speederpad.
So you might want to investigate it further with us on our [Discord](https://discord.gg/mainsail) as this shouldn't happen.

**Comment by brdi** [comment]
@mryel00I
had the same problem with an X701 ProXtend  and after reading this thread i inserted
the following lines after the Parse Arg section in the crowsnest file using SSH on MobaXterm on my windows PC.

  ## _ start SOL1: Solving camera not connecting isssue_____________________________
  if (( $(awk '{print $1}' /proc/uptime | cut -d '.' -f 1) < 160 )); then
      sleep 5
  fi
  ## _ end SOL1 ____________________________

Worked like a charm. Camera was there directly after starting the system. :-)

However, this morning (march 4th 2024) on the  Speeder Pad an update came along.
The crowsnest update failed :(.
Being a newbie, I have my FLSUN V400 for barely 3 weeks now, I removed the lines I added and started again. 
The crownest Update still is invalid.
Using Moonsail I cleaned the dirty file "softly" 
After that, the updat was OK, but again: No camera connected :-(
So I'll put in the lines again...

Hoped this issue would be solved in a new releas. Did I miss something?

**Comment by brdi** [comment]
@mryel00I
had the same problem with an X701 ProXtend and after reading this thread i inserted
the following lines after the Parse Arg section in the crowsnest file using SSH on MobaXterm on my windows PC.

##  start SOL1: Solving camera not connecting isssue_____________________________

if (( $(awk '{print $1}' /proc/uptime | cut -d '.' -f 1) < 160 )); then
sleep 5
fi

## end SOL1 ____________________________

Worked like a charm. Camera was there directly after starting the system. :-)
However, this morning (march 4th 2024) on the Speeder Pad an update came along.
The crowsnest update failed :(.
Being a newbie, I have my FLSUN V400 for barely 3 weeks now, I removed the lines I added and started again.
The crownest Update still is invalid.
Using Moonsail I cleaned the dirty file "softly"
After that, the update was OK, but again: No camera connected :-(
But after closing down everything and restaring it looks to work perfectly without the code added.

**Comment by mryel00** [comment]
@brdi Same question as in the comment directly above yours:
> Did you read the last two comments from me on this post?

That fix that you are applying manually is already in Crowsnest since 15.11.2023. I hided the comments for a good reason, because they are outdated but I guess I have to edit at least everything I wrote earlier that people don't edit code in that's already there. If you edit the Crowsnest code you can ofc not update anymore, because you got changes that git cannot handle on it's own.
I gave the whole instructions on what to do in [this comment](https://github.com/mainsail-crew/crowsnest/issues/117#issuecomment-1806943132).

**Comment by shubenets** [comment]
Have the same issue, cam working fine only after reboot. I chek log and find error that ustreamer cannot find cam on /dev/video0/ After manual reboot crowsnets, ustreamer search and correctly find cam by-id that i was insert in my crowsnest.conf. After manual reboot all working fine, i was try to change port and add delay to CROWSNEST_ARGS with no result.

[crowsnest (3).log](https://github.com/mainsail-crew/crowsnest/files/14817553/crowsnest.3.log)
[crowsnest.conf.txt](https://github.com/mainsail-crew/crowsnest/files/14817556/crowsnest.conf.txt)

**Comment by mryel00** [comment]
> I chek log and find error that ustreamer cannot find cam on /dev/video0/

You sent a file with 14k lines. I cannot find the thing you are talking about. It's always finding the device and then ustreamer fails. Why it fails? There is no reason printed inside the log, but it's most likely due to the issue of either the port is being blocked or network not started yet, because ustreamer doesn't exits if the device wasn't found.

> i was try to change port

So it's most likely network not there yet.

> add delay to CROWSNEST_ARGS with no result.

Maybe increase it? It's an Intel CPU with some Bookworm Debian. No idea if there exists some similar issue.

**_BUT this isn't the place for support. Join our [Discord](https://discord.gg/mainsail) or our [GitHub Discussions](https://github.com/orgs/mainsail-crew/discussions) to get help._**

To investigate on this yourself, you should first enable debug log and delete your log at least once....
